### PR TITLE
Update typescript

### PR DESCRIPTION
**Pull request by chicoxyzzy**
DO NOT MERGE.
see https://github.com/facebook/react/pull/5485#issuecomment-216410166 and https://github.com/facebook/react/pull/5485#discussion_r45021573

### CODE REVIEW

**Comment by chicoxyzzy** [comment]
Oops I forgot to remove `noImplicitAny` options for now

**Comment by facebook-github-bot-9** [comment]
@chicoxyzzy updated the pull request.

**Comment by chicoxyzzy** [comment]
Probably tests pass even with invalid definition because `jest.d.ts` is not included in compilation process

**Comment by chicoxyzzy** [comment]
What about usage of `gulp-typescript` if it will help? Configuration could be easier but we'll need to change npm test script to run TypeScript Compiler and then Jest

File: scripts/jest/ts-preprocessor.js [hunk 1]
```diff
@@ -3,13 +3,16 @@
 var fs = require('fs');
 var path = require('path');
 var ts = require('typescript');
+var os = require('os');
 
-var tsOptions = {module: 'commonjs'};
+var tsOptions = {
+  module: 'commonjs',
```
Comment (RReverser) [inline scripts/jest/ts-preprocessor.js]: You need to use `ModuleKind`, not a `string` now: https://github.com/Microsoft/TypeScript/blob/master/src/compiler/types.ts#L2456 + https://github.com/Microsoft/TypeScript/blob/master/src/compiler/types.ts#L2535-L2543

**Comment by RReverser** [comment]
There is another interesting compiler option - `diagnostics` - https://github.com/Microsoft/TypeScript/blob/master/src/compiler/types.ts#L2444 - not sure if it will help, but might be worth trying to turn it on.

**Comment by chicoxyzzy** [comment]
unfortunately it doesn't help

**Comment by facebook-github-bot-0** [comment]
@chicoxyzzy updated the pull request.

File: scripts/jest/ts-preprocessor.js [hunk 2]
```diff
@@ -3,13 +3,17 @@
 var fs = require('fs');
 var path = require('path');
 var ts = require('typescript');
+var os = require('os');
 
-var tsOptions = {module: 'commonjs'};
+var tsOptions = {
+  module: ts.ModuleKind.CommonJS,
+  jsx: 'React',
```
Comment (DanielRosenwasser) [inline scripts/jest/ts-preprocessor.js]: I believe this should be `ts.JsxEmit.React`

Comment (chicoxyzzy) [inline scripts/jest/ts-preprocessor.js]: Oops. You are right thank you!

**Comment by facebook-github-bot-9** [comment]
@chicoxyzzy updated the pull request.

File: scripts/jest/ts-preprocessor.js [hunk 3]
```diff
@@ -55,34 +59,37 @@ function compile(content, contentFilename) {
...
-    getNewLine: function() {
+    getNewLine() {
       return '\n';
     },
+    fileExists(filename) {
+      return ts.sys.fileExists(filename);
+    },
+    useCaseSensitiveFileNames() {
+      var platform = os.platform();
+      return platform !== 'win32' && platform !== 'win64' && platform !== 'darwin';
```
Comment (zpao) [inline scripts/jest/ts-preprocessor.js]: Not perfect since it's not a 1:1 mapping to filesystem but good enough :)

**Comment by chicoxyzzy** [comment]
Maybe we can just take `gulp-typescript` package and run gulp task by `npm test`? This task can run typescript and then run jest. Don't know it will work but I don't know how to make current changes work proper as well :)

File: scripts/jest/ts-preprocessor.js (hunk 3 above)
Comment (chicoxyzzy) [inline scripts/jest/ts-preprocessor.js]: improved

**Comment by facebook-github-bot-0** [comment]
@chicoxyzzy updated the pull request.

**Comment by chicoxyzzy** [comment]
@dmitriiabramov do you know how I can update my PR to try https://github.com/facebook/react/pull/6686#issuecomment-216668137

**Comment by DanielRosenwasser** [comment]
Do you mean using Gulp with TypeScript in general? We have a [tutorial if you'd like](https://www.typescriptlang.org/docs/handbook/gulp.html).

Can I understand what the specific problems you're running into are? I'm having trouble getting context but I'd like to help if I can.

**Comment by chicoxyzzy** [comment]
@DanielRosenwasser the idea is to use `gulp-typescript` package (which has preconfigured TS compiler). I thought there can be `test` npm script which can run `test` gulp task which can look like so (pseudocode):

```
var tsc = require('gulp-typescript');
var tsProject = tsc.createProject('tsconfig.json');

gulp.src(TEST_FILES)
    .pipe(tsc(tsProject))
    .pipe(jest)
    .pipe(gulp.dest(DEST));
```

The problem here is I don't understand what `.pipe(jest)` is because now jest is preconfigured using `package.json` and `test` npm script now is just `"test": "jest"`.

Maybe all this won't help at all. Main problem now is https://github.com/facebook/react/pull/5485#discussion_r45021573

**Comment by zpao** [comment]
I don't think trying to involve gulp here will be beneficial. Learning from the compilation options that are happening there might be valuable but not the rest of the gulp stuff since we're doing compilation in the jest preprocessor.

**Comment by facebook-github-bot-7** [comment]
@chicoxyzzy updated the pull request.

**Comment by chicoxyzzy** [comment]
closed in favour of #6804

**Pull request closed without merging**